### PR TITLE
HTTP server and Rest Api

### PR DESCRIPTION
**Pull request by MostafaGomaa**
This PR implements an HTTP server and a rest-api processor for the refbox. (on port 8088)
Then it implements handlers and processors for CLIPS environment, allowing access to clips facts by Get queries.

Usage:
`localhost:8088/clips/facts `
reply with all facts in a generic  JSON representation for Facts
`localhost:8088/clips/facts/{tmpl-name}?slot1=value1,.. `
Return facts with tmpl-name matching the query args.  the reply is a generic JSON representation for Facts
`localhost:8088/clips/{TemplateModel}s?slot1=value1,.. `
Reply with a JSON build from the generated models of the specific fact templates. (i.e, the JSON structured according to the predefined models ).

Currently, models parsing clips templates are implemented for and queriable by
.localhost:8088/clips/orders
localhost:8088/clips/machines
localhost:8088/clips/ring-spec
localhost:8088/clips/gamestate
localhost:8088/clips/robot

(Disclaimer: This is highly adapted from and extending on  fawkes::webview plugin, aspects and library)

### CODE REVIEW

**Comment by morxa** [comment]
As discussed, this PR is still work in progress and does not need a review at the moment.

**Comment by morxa** [comment]
For fixing the build, you probably just need to install additional packages in the circleci config.

**Comment by morxa** [comment]
@MostafaGomaa can you please fix the build? We should get this merge-ready.

**Comment by MostafaGomaa** [comment]
@morxa Thanks for you patience, this is now ready for review !

**Comment by MostafaGomaa** [comment]
Seems the docker build is failing but I do not have authentication to access the page.
@morxa can you please give access or advice on the matter
 My docker id: mostafagomaa

**Comment by morxa** [comment]
Done, you should now have access.

The problem seems to be that `libmicrohttpd` is missing in the Dockerfile:
```
 --> [0;31mOmitting Webview Library[0;39m (libmicrohttpd not found)
--> [0;31mWarnings are considered to be errors![0;39m Build with FAIL_ON_WARNING=0 to disable.
```

**Comment by MostafaGomaa** [comment]
Strang..I it still give me 404 error .circleci/config.yaml and is being built successfully. are those configs not used to build the docker image as well?

**Comment by morxa** [comment]
No, Docker has its own list of packages in the Dockerfile.

I also get a 404, I think the link is broken. But you can browse it via hub.docker.com:
https://hub.docker.com/repository/docker/robocuplogistics/rcll-refbox/builds

**Comment by MostafaGomaa** [comment]
Finally passed the docker build. Ready for your review @morxa  :)

File: src/libs/rest_api/service_browse_handler.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	                           const socklen_t         addr_size,
+	                           uint16_t                port,
+	                           std::list<std::string> &txt,
+	                           int                     flags);
+	virtual void service_removed(const char *name, const char *type, const char *domain);
+
+	/** A map of services.
+   * Maps service names to NetworkService instances describing the service
+   * in more detail. */
+	typedef std::map<std::string, fawkes::NetworkService *> ServiceList;
```
Comment (morxa) [inline src/libs/rest_api/service_browse_handler.h:59]: Can we get rid of this raw pointer in the typedef? There are multiple options:
1. store as member, not as pointer: `std::map<std::string, fawkes::NetworkService>`
1. as unique ptr: `std::map<std::string, std::unique_ptr<fawkes::NetworkService>>`
1. as shared ptr: `std::map<std::string, std::shared_ptr<fawkes::NetworkService>>`

The first option would be best, but I'm not sure if it's possible, as copying a `NetworkService` is probably not possible.

File: src/libs/rest_api/webview_server.cpp [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+	nnresolver_        = nnresolver;
+	service_publisher_ = service_publisher;
+	service_browser_   = service_browser;
+	config_            = config;
+	logger_            = logger;
+
+	url_manager_      = new WebUrlManager();
+	request_manager_  = new WebRequestManager();
+	rest_api_manager_ = new WebviewRestApiManager();
+	clips_rest_api_   = new ClipsRestApi(env, clips_mutex, rest_api_manager_, logger_);
```
Comment (morxa) [inline src/libs/rest_api/webview_server.cpp:89]: Also here, I don't think any of those need to be raw pointers.

File: src/libs/rest_api/webview_server.cpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+WebviewServer::~WebviewServer()
+{
+}
+
+void
+WebviewServer::init()
```
Comment (morxa) [inline src/libs/rest_api/webview_server.cpp:97]: Move to constructor if possible (should be possible here).

File: src/libs/rest_api/webview_server.cpp [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+		//Clips Rest api
+		clips_rest_api_->init();
+		clips_rest_api_->start();
+	} catch (Exception &e) {
+		throw;
+	}
+}
+
+void
+WebviewServer::finalize()
```
Comment (morxa) [inline src/libs/rest_api/webview_server.cpp:190]: Move to destructor.

File: src/refbox/refbox.h [hunk 5]
```diff
@@ -216,10 +218,14 @@ class LLSFRefBox
...
-	std::unique_ptr<fawkes::NetworkNameResolver> nnresolver_;
-	std::unique_ptr<fawkes::NetworkService>      refbox_service_;
+	fawkes::AvahiThread                     avahi_thread_;
+	std::unique_ptr<fawkes::NetworkService> refbox_service_;
 #endif
+	std::unique_ptr<WebviewServer> rest_api_thread_;
+
+	std::unique_ptr<fawkes::NetworkNameResolver> nnresolver_;
+	fawkes::ServicePublisher *                   service_publisher_;
+	fawkes::ServiceBrowser *                     service_browser_;
```
Comment (morxa) [inline src/refbox/refbox.h:228]: Also here, please avoid raw pointers.

**Comment by morxa** [comment]
@MostafaGomaa any updates?

It looks like this needs a rebase again.

File: .circleci/config.yml [hunk 6]
```diff
@@ -29,7 +29,10 @@ jobs:
               protobuf-devel \
               which \
               yaml-cpp-devel \
-              yamllint
+              yamllint \
+              libmicrohttpd-devel \
+              rapidjson-devel \
+              apr-util-devel
```
Comment (morxa) [inline .circleci/config.yml:35]: Please keep the list sorted.

**Comment by morxa** [comment]
Also I noticed that the CPU is at 100% with this branch, can you check if this the same for you and if this is actually caused by changes here?

File: cfg/config.yaml [hunk 7]
```diff
@@ -205,3 +205,39 @@ llsfrb:
...
   webshop:
     enable: false
+
+webview:
+  # TCP port for Webview HTTP requests; TCP port
+  port: !tcp-port 8088
+
+  # True to allow client side caching. Typically you want this disabled
+  # to force proper reloads each time with fresh data.
+  client_side_caching: true
```
[the remark this answers is not in the document]
Comment (MostafaGomaa) [reply cfg/config.yaml]: Done!

File: cfg/config.yaml [hunk 8]
```diff
@@ -205,3 +205,39 @@ llsfrb:
...
+      all: true
+      # Allow access only from specific domains
+      # origins: ["http://localhost:4200"]
+
+    # time to allow clients to cache our response
+    max-age: 86400
+
+  # libmicrohttpd thread pool settings
+  # Using a thread pool allows for the concurrent processing of
+  # incoming requests and should generally be enabled.
```
[the remark this answers is not in the document]
Comment (MostafaGomaa) [reply cfg/config.yaml:246]: Done!

File: src/libs/rest_api/Makefile [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    WARN_TARGETS += warning_rapidjson
+  endif
+  ifneq ($(HAVE_BOOST_LIBS),1)
+    WARN_TARGETS_BOOST = $(foreach l,$(REQ_BOOST_LIBS),$(if $(call boost-have-lib,$l),, warning_boost_$l))
+  endif
+endif
+
+ifeq ($(OBJSSUBMAKE),1)
+  ifneq ($(WARN_TARGETS),)
+all: $(WARN_TARGETS)
```
[the remark this answers is not in the document]
Comment (MostafaGomaa) [reply src/libs/rest_api/Makefile]: Done!

File: src/libs/rest_api/Makefile [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+  ifneq ($(HAVE_RAPIDJSON),1)
+    WARN_TARGETS += warning_rapidjson
+  endif
+  ifneq ($(HAVE_BOOST_LIBS),1)
+    WARN_TARGETS_BOOST = $(foreach l,$(REQ_BOOST_LIBS),$(if $(call boost-have-lib,$l),, warning_boost_$l))
+  endif
+endif
+
+ifeq ($(OBJSSUBMAKE),1)
+  ifneq ($(WARN_TARGETS),)
```
[the remark this answers is not in the document]
Comment (MostafaGomaa) [reply src/libs/rest_api/Makefile]: Done!

File: src/libs/clips_rest_api/clips-rest-api.cpp [hunk 11]
```diff
@@ -0,0 +1,530 @@
...
+	                                                   std::bind(&ClipsRestApi::cb_get_ring_spec,
+	                                                             this,
+	                                                             std::placeholders::_1));
+	rest_api_->add_handler<WebviewRestArray<Points>>(WebRequest::METHOD_GET,
+	                                                 "/points",
+	                                                 std::bind(&ClipsRestApi::cb_get_points,
+	                                                           this,
+	                                                           std::placeholders::_1));
+
+	webview_rest_api_manager_->register_api(rest_api_);
```
[the remark this answers is not in the document]
Comment (MostafaGomaa) [reply src/libs/clips_rest_api/clips-rest-api.cpp]: Done, also some refactoring to the clips-rest-api along the way
 - Away with rest_api and rest_api_manager members
 - Extends RestApi base class instead
 - Do not extend Thread

File: src/libs/clips_rest_api/clips-rest-api.cpp [hunk 12]
```diff
@@ -0,0 +1,530 @@
...
+/** Destructor. */
+ClipsRestApi::~ClipsRestApi()
+{
+}
+
+void
+ClipsRestApi::init()
+{
+	logger_->log_info("ClipsRestApi", "Initializing thread");
+	rest_api_ = new WebviewRestApi("clips", logger_);
```
[the remark this answers is not in the document]
Comment (MostafaGomaa) [reply src/libs/clips_rest_api/clips-rest-api.cpp]: Refactored away, now ClipsRestApi extends WebviewRestApi.

File: src/libs/rest_api/webview_server.cpp (hunk 2 above)
Comment (MostafaGomaa) [reply src/libs/rest_api/webview_server.cpp]: Done, Also removed all pointers in libs\rest-api

File: .circleci/config.yml (hunk 6 above)
Comment (MostafaGomaa) [reply .circleci/config.yml]: Done!

**Comment by MostafaGomaa** [comment]
Done, I have applied your comments and did some refactoring on the way
@morxa please take another look :)

**Comment by MostafaGomaa** [comment]
> Looks like the formatting is broken, have you run `make quick-check`? (The CI test was broken, fixed in #77).

True, I forgot to check the formating. Done :)

> Other than that, this looks good!

**Comment by morxa** [comment]
Thanks for working on this!

**Pull request merged**